IPython/utils/traitlets.py

```diff
@@ -570,43 +572,73 @@ def __new__(cls, *args, **kw):
+        inst._cross_validation_lock = False
         return inst
 
     def __init__(self, *args, **kw):
         # Allow trait values to be set using keyword arguments.
         # We need to use setattr for this to trigger validation and
         # notifications.
-        
         with self.hold_trait_notifications():
             for key, value in iteritems(kw):
                 setattr(self, key, value)
-    
+
     @contextlib.contextmanager
     def hold_trait_notifications(self):
-        """Context manager for bundling trait change notifications
-        
-        Use this when doing multiple trait assignments (init, config),
-        to avoid race conditions in trait notifiers requesting other trait values.
+        """Context manager for bundling trait change notifications and cross
+        validation.
+
+        Use this when doing multiple trait assignments (init, config), to avoid
+        race conditions in trait notifiers requesting other trait values.
         All trait notifications will fire after all values have been assigned.
         """
-        _notify_trait = self._notify_trait
-        notifications = []
-        self._notify_trait = lambda *a: notifications.append(a)
-        
-        try:
+        if self._cross_validation_lock is True:
             yield
-        finally:
-            self._notify_trait = _notify_trait
-            if isinstance(_notify_trait, types.MethodType):
-                # FIXME: remove when support is bumped to 3.4.
-                # when original method is restored,
-                # remove the redundant value from __dict__
-                # (only used to preserve pickleability on Python < 3.4)
-                self.__dict__.pop('_notify_trait', None)
-        # trigger delayed notifications
-        for args in notifications:
-            self._notify_trait(*args)
+        else:
+            self._cross_validation_lock = True
+            cache = {}
+            notifications = {}
+            _notify_trait = self._notify_trait
+
+            def cache_values(*a):
+                cache[a[0]] = a
+
+            def hold_notifications(*a):
+                notifications[a[0]] = a
+
+            self._notify_trait = cache_values
+
+            try:
+                yield
+            finally:
+                try:
+                    self._notify_trait = hold_notifications
+                    for name in cache:
+                        if hasattr(self, '_%s_validate' % name):
+                            cross_validate = getattr(self, '_%s_validate' % name)
+                            setattr(self, name, cross_validate(getattr(self, name), self))
+                except TraitError as e:
+                    self._notify_trait = lambda *x: None
+                    for name in cache:
+                        if cache[name][1] is not Undefined:
+                            setattr(self, name, cache[name][1])
+                        else:
+                            delattr(self, name)
```

Comment:
Can you make sure to add a test for the rejected validation? I haven't seen delattr on a traitlet before, and I'd like to be sure it behaves as expected:

```python
c = Test()
with c.hold_trait_notifications():
    c.trait = invalid_value

assert c.trait == default_value
```

It might be that this should be a `self._trait_values.pop(name)`